### PR TITLE
Refactor builtins

### PR DESCRIPTION
**Pull request by notspiff**
- splits the monolith file in multiple files
- adds an registry infrastructure to register commands from various files/classes
- tries to doxy all functions (some of them have really funk calling conventions)

need project updates.

### CODE REVIEW

**Comment by wsnipex** [comment]
great stuff! I've noted only 2 weeks ago that builtins need a refactor badly :)

File: xbmc/interfaces/builtins/ApplicationBuiltins.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+{
+  g_application.getNetwork().WakeOnLan((char*)params[0].c_str());
+
+  return 0;
+}
+
+CBuiltins::CommandMap CApplicationBuiltins::GetOperations() const
+{
+  return {
+           {"extract", {"Extracts the specified archive", 1, Extract}},
```
Comment (da-anda) [inline xbmc/interfaces/builtins/ApplicationBuiltins.cpp]: minor nitpick: mind using the same nice idention of the second value like you did in all the other files?

Comment (notspiff) [inline xbmc/interfaces/builtins/ApplicationBuiltins.cpp]: done.

**Comment by Paxxi** [comment]
I really don't approve of the InputManager implementation, why create global methods and a copy of the CRemoteControl member? Stick the methods as static method into InputManager and use the Get() method to get the instance, since you're a member of InputManager you can access the private members directly if needed.

thumbs up for the effort though, will poke and prod at it :)

**Comment by notspiff** [comment]
Fine. I loath class statics in singletons, in particular if the static use the singleton acessor. But if you insist..

**Comment by notspiff** [comment]
And just because right should be right; there are no globals involved.

**Comment by Paxxi** [comment]
sorry, didn't mean globals as in variables but static functions in the global scope.

Have you thought about using std::function instead of a raw function pointer? std::function can be used for member functions so you could bundle things up more neatly instead of having all functions in the global namespace you can stick them into the classes with the GetOperations

**Comment by notspiff** [comment]
Read up on static functions, they are not in the global namespace. No i didnt consider that as i see no gain.

**Comment by Paxxi** [comment]
yup, I was wrong about that, thanks for pointing it out :)

**Comment by Montellese** [comment]
I haven't gone through the whole PR yet (actually only the first two commits).

Personally I prefer to use interfaces (with explicitly defined methods) over templates (with implicitly defined methods) for such things. That way it would also be possible to register builtins from other places without having to explicitly know about them and add them to `Builtins.cpp`.
Apart from that it probably doesn't have any other advantages so I'm not gonna insist on it if everyone else thinks this is good enough.

Either way, great work and thanks for looking into this.

**Comment by notspiff** [comment]
i prefer compiler errors to subtle runtime errors, so there we have different opinions. not that it matter much in this case either way. other concerns addressed.

**Comment by notspiff** [comment]
as for allowing external register calls, there is no problem making the template public, so that by itself doesn't require an interface. however, if you do that, you need locking and all the fun that entails and i want to keep it simple right now.

**Pull request closed without merging**